python/lsst/daf/butler/_compat.py

```diff
@@ -1,272 +0,0 @@
-# This file is part of pipe_base.
-#
-# Developed for the LSST Data Management System.
-# This product includes software developed by the LSST Project
-# (https://www.lsst.org).
-# See the COPYRIGHT file at the top-level directory of this distribution
-# for details of code ownership.
-#
-# This software is dual licensed under the GNU General Public License and also
-# under a 3-clause BSD license. Recipients may choose which of these licenses
-# to use; please see the files gpl-3.0.txt and/or bsd_license.txt,
-# respectively.  If you choose the GPL option then the following text applies
-# (but note that there is still no warranty even if you opt for BSD instead):
-#
-# This program is free software: you can redistribute it and/or modify
-# it under the terms of the GNU General Public License as published by
-# the Free Software Foundation, either version 3 of the License, or
-# (at your option) any later version.
-#
-# This program is distributed in the hope that it will be useful,
-# but WITHOUT ANY WARRANTY; without even the implied warranty of
-# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-# GNU General Public License for more details.
-#
-# You should have received a copy of the GNU General Public License
-# along with this program.  If not, see <https://www.gnu.org/licenses/>.
-
-"""Code to support backwards compatibility."""
-
-from __future__ import annotations
-
-__all__ = ["PYDANTIC_V2", "_BaseModelCompat"]
-
-import sys
-from collections.abc import Callable
-from typing import TYPE_CHECKING, Any, Literal
-
-from pydantic import BaseModel
-from pydantic.fields import FieldInfo
-from pydantic.version import VERSION as PYDANTIC_VERSION
-
-if sys.version_info >= (3, 11, 0):
-    from typing import Self
-else:
-    from typing import TypeVar
-
-    Self = TypeVar("Self", bound="_BaseModelCompat")  # type: ignore
-
-
-PYDANTIC_V2 = PYDANTIC_VERSION.startswith("2.")
-
-# This matches the pydantic v2 internal definition.
-IncEx = set[int] | set[str] | dict[int, Any] | dict[str, Any] | None
-
-if PYDANTIC_V2:
-
-    class _BaseModelCompat(BaseModel):
-        """Methods from pydantic v1 that we want to emulate in v2.
-
-        Some of these methods are provided by v2 but issue deprecation
-        warnings.  We need to decide whether we are also okay with deprecating
-        them or want to support them without the deprecation message.
-        """
-
-        def json(
```

Comment:
I wasn't really sure if we were going to break people by dropping the `json` method here. I think that there might be people who use `SerializedDatasetRef` outside of our middleware but maybe only `to_simple` and `from_simple` rather than calling `.json` and `.parse_obj`. Maybe @mfisherlevine ?

Reply:
I've been thinking of all the `Serialized*` as middleware-internal, but I guess that's not quite true since those `*_simple` methods are public.

I'd really rather not go through a deprecation period for an interface we just inherit from a third-party package, but I'm happy to stall merging this for a few days to see if Merlin chimes in.

Reply:
The compromise is to merge this tomorrow after the next weekly and then you have a week to find out from people if we need to put a shim back in specifically for `SerializedDatasetRef`.